src/python/grpcio_tests/tests/health_check/_health_servicer_test.py

```diff
@@ -38,29 +38,170 @@ def _consume_responses(response_iterator, response_queue):
+            self.assertEqual(health_pb2.HealthCheckResponse.SERVING,
+                             response1.status)
+            self.assertEqual(health_pb2.HealthCheckResponse.SERVING,
+                             response2.status)
+
+            rendezvous1.cancel()
+            rendezvous2.cancel()
+            thread1.join()
+            thread2.join()
+            self.assertTrue(response_queue1.empty())
+            self.assertTrue(response_queue2.empty())
+
+        def test_cancelled_watch_removed_from_watch_list(self):
+            request = health_pb2.HealthCheckRequest(service=_WATCH_SERVICE)
+            response_queue = queue.Queue()
+            rendezvous = self._stub.Watch(request)
+            thread = threading.Thread(
+                target=_consume_responses, args=(rendezvous, response_queue))
+            thread.start()
+
+            response = response_queue.get(timeout=test_constants.SHORT_TIMEOUT)
+            self.assertEqual(health_pb2.HealthCheckResponse.SERVICE_UNKNOWN,
+                             response.status)
+
+            rendezvous.cancel()
+            self._servicer.set(_WATCH_SERVICE,
+                               health_pb2.HealthCheckResponse.SERVING)
+            thread.join()
+
+            # Wait, if necessary, for serving thread to process client cancellation
+            timeout = time.time() + test_constants.SHORT_TIMEOUT
+            while time.time(
+            ) < timeout and self._servicer._on_next_callbacks[_WATCH_SERVICE]:
+                time.sleep(1)
+            self.assertFalse(self._servicer._on_next_callbacks[_WATCH_SERVICE],
+                             'watch set should be empty')
+            self.assertTrue(response_queue.empty())
+
+
+class HealthServicerTest(BaseWatchTests.WatchTests):
 
     def setUp(self):
-        self._servicer = health.HealthServicer()
-        self._servicer.set('', health_pb2.HealthCheckResponse.SERVING)
-        self._servicer.set(_SERVING_SERVICE,
-                           health_pb2.HealthCheckResponse.SERVING)
-        self._servicer.set(_UNKNOWN_SERVICE,
-                           health_pb2.HealthCheckResponse.UNKNOWN)
-        self._servicer.set(_NOT_SERVING_SERVICE,
-                           health_pb2.HealthCheckResponse.NOT_SERVING)
-        self._server = test_common.test_server()
-        port = self._server.add_insecure_port('[::]:0')
-        health_pb2_grpc.add_HealthServicer_to_server(self._servicer,
-                                                     self._server)
-        self._server.start()
-
-        self._channel = grpc.insecure_channel('localhost:%d' % port)
-        self._stub = health_pb2_grpc.HealthStub(self._channel)
-
-    def tearDown(self):
-        self._server.stop(None)
-        self._channel.close()
+        super(HealthServicerTest, self).start_server(
+            health.HealthServicer(
+                experimental_non_blocking=False, experimental_thread_pool=None))
```

Comment:
Do you mean `experimental_non_blocking=True`?
Do we need to test for the `experimental_thread_pool`?

Reply:
Oops, most definitely. Thanks for catching, fixed.